src/themis/sym_enc_message.c

```diff
@@ -176,303 +177,371 @@ themis_status_t themis_sym_plain_decrypt(uint32_t alg,
+                                                             const uint8_t* key,
+                                                             size_t key_length,
+                                                             const uint8_t* user_context,
+                                                             size_t user_context_length,
+                                                             uint8_t* derived_key,
+                                                             size_t* derived_key_length)
+{
+    size_t required_length = soter_alg_key_length(hdr->alg);
+    switch (required_length) {
+    case SOTER_SYM_256_KEY_LENGTH / 8:
+    case SOTER_SYM_192_KEY_LENGTH / 8:
+    case SOTER_SYM_128_KEY_LENGTH / 8:
+        break;
+    default:
+        return THEMIS_FAIL;
+    }
+    /* Internal buffer must have suitable size */
+    if (*derived_key_length < required_length) {
+        return THEMIS_FAIL;
+    }
+    *derived_key_length = required_length;
+    /*
+     * SOTER_SYM_NOKDF means Soter KDF in this context.
+     * This is the only KDF allowed for master key API.
+     */
+    switch (soter_alg_kdf(hdr->alg)) {
+    case SOTER_SYM_NOKDF: {
+        uint8_t message_length_label[sizeof(uint32_t)] = {0};
+        stream_write_uint32LE(message_length_label, hdr->message_length);
+        return themis_sym_kdf(key,
+                              key_length,
+                              THEMIS_SYM_KDF_KEY_LABEL,
+                              message_length_label,
+                              sizeof(message_length_label),
+                              user_context,
+                              user_context_length,
+                              derived_key,
+                              *derived_key_length);
+    }
+    default:
+        return THEMIS_FAIL;
+    }
+}
+
+#ifdef SCELL_COMPAT
+static themis_status_t themis_auth_sym_derive_encryption_key_compat(
+    const struct themis_scell_auth_token_key* hdr,
+    const uint8_t* key,
+    size_t key_length,
+    const uint8_t* user_context,
+    size_t user_context_length,
+    uint8_t* derived_key,
+    size_t* derived_key_length)
+{
+    size_t required_length = soter_alg_key_length(hdr->alg);
+    switch (required_length) {
+    case SOTER_SYM_256_KEY_LENGTH / 8:
+    case SOTER_SYM_192_KEY_LENGTH / 8:
+    case SOTER_SYM_128_KEY_LENGTH / 8:
+        break;
+    default:
+        return THEMIS_FAIL;
+    }
+    /* Internal buffer must have suitable size */
+    if (*derived_key_length < required_length) {
+        return THEMIS_FAIL;
+    }
+    *derived_key_length = required_length;
+    /*
+     * SOTER_SYM_NOKDF means Soter KDF in this context.
+     * This is the only KDF allowed for master key API.
+     */
+    switch (soter_alg_kdf(hdr->alg)) {
+    case SOTER_SYM_NOKDF: {
+        /*
+         * Themis 0.9.6 incorrectly used 64-bit size_t (as uint64_) for this field.
```

Comment:
> why not just wrap this small piece of code with #define instead copy/paste 70% of code?

I vaguely recall there was a reason but I can’t remember it right now. I’ll see what if this can be cleaned up. Maybe it has something to do with the structure of code.

While we’re here, I’d like @vixentael to chime in for a second opinion. How about keeping this compatibility shim enabled permanently, removing all this `#ifdef` magic around it, and making `NO_SCELL_COMPAT` a no-op? I’ve been thinking about it for some time and can’t find any reasons why we would like to have this as opt-out. “Hey, psst, user! Do want to try turning this switch off? It will not improve performance on correct data, but can make some old data to be undecryptable. Wanna try?”

Reply:
Okay, I’ve extracted the varying parts into `themis_auth_sym_kdf_context()` and `themis_auth_sym_kdf_context_compat()`. Now there is less code duplication. Ironically, this is longer by 5 lines because of error handling (which is going to be optimized out).